Trainer will now fold and log estimated norm scaling factor

curt-tigges opened the pull request:
# Description

Addresses https://github.com/jbloomAus/SAELens/issues/222. After performing a fit, the trainer will fold estimated_norm_scaling_factor into the SAE weights and will log this value to Weights and Biases.

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update



# Checklist:

- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] I have not rewritten tests relating to key interfaces which would affect backward compatibility

<!--
As you go through the checklist above, you can mark something as done by putting an x character in it

For example,
- [x] I have done this task
- [ ] I have not done this task
-->

### You have tested formatting, typing and unit tests (acceptance tests not currently in use)

- [x] I have run `make check-ci` to check format and linting. (you can run `make format` to format code if needed.)

jbloomAus commented:
@curt-tigges I think it would be good to have code in the SAE Loader for pretrained saes such as Tom McGrath's GPT2 small residual stream SAEs where if we save the estimated scaling factor once to the yaml, then we don't need to re-estimate it before using the SAE (as we do [here](https://github.com/jbloomAus/SAELens/blob/db89dbc813e5a64a7609165474e3c66e2b537925/tests/benchmark/test_eval_all_loadable_saes.py#L89). 

To do:

- [ ] Remove change to wandb logging.
- [ ] Ensure we don't modify SAE during training to save with the factor folded in. 
- [ ] Add estimated cfg to yaml so an SAE saved with "expected_average_only_in" has it folded in and set to None when loading from pretrained.

curt-tigges commented:
@jbloomAus Rebased and made the changes you suggested. Value will not be logged in wandb. Scaling factor is folded in only after training, when saving the last checkpoint.

In addition, I did the following:
- Precomputed norm_scaling_factor for all SAEs with "normalize_activations" == "expected_average_only_in" and added it to the YAML
- Added a function to search the YAML for these values if needed
- Modified from_pretrained to identify if "normalize_activations" == "expected_average_only_in" in the config, and if so, executes the above function to find the norm_scaling_factor. It will then:
    - Fold norm_scaling_factor into the weights
    - Set cfg_dict["normalize_activations"] = "none"

jbloomAus commented:
Thanks!
